apps/prairielearn/assets/scripts/instructorAssessmentInstancesClient.tsx

```diff
@@ -401,93 +405,325 @@ onDocumentReady(() => {
+    } else if (plusMinus === '-1') {
+      explanation = 'This will subtract the given amount of time from the remaining time limit.';
+    } else if (plusMinus === 'unlimited') {
+      explanation = 'This will remove the time limit and the assessment will remain open.';
+    } else if (plusMinus === 'expire') {
+      explanation = 'This will expire the time limit and the assessment will be closed.';
+    }
+    return <small class="text-muted">{explanation}</small>;
+  }
+
+  function ReopenClosedCheckbox({
+    handleChange,
+    reopenClosed,
+  }: {
+    handleChange: (e: Event) => void;
+    reopenClosed: string;
+  }) {
+    return (
+      <div class="form-check">
+        <input
+          class="form-check-input"
+          type="checkbox"
+          name="reopen_closed"
+          value={reopenClosed}
+          id="reopen_closed"
+          onChange={(e) => handleChange(e)}
+        />
+        <label class="form-check-label" for="reopen_closed">
+          Also re-open closed instances
+        </label>
+      </div>
+    );
+  }
+
+  function TimeLimitEditPopover({
+    row,
+  }: {
+    row: {
+      action: string;
+      assessment_instance_id: number;
+      date: string;
+      has_closed_instance: boolean;
+      has_open_instance: boolean;
+      total_time: string;
+      time_remaining: string;
+      time_remaining_sec: number;
+      open: boolean;
+    };
+  }) {
+    // Because we need to update values on the Preact virtual DOM, we need to manage our
+    // form values as state. These can then be passed and used by the child components to
+    // display values or used for conditional rendering.
+    const [form, setForm] = useState({
+      assessment_instance_id: row.assessment_instance_id ?? null,
+      plus_minus: '+1',
+      time_add: 5,
+      date: Temporal.Now.zonedDateTimeISO(timezone).toPlainDateTime().toString().slice(0, 16),
+      reopen_closed: 'false',
+      reopen_without_limit: 'true',
+    });
+    const [showTimeLimitOptions, setShowTimeLimitOptions] = useState(
+      row.action === 'set_time_limit_all' || row.open || form.reopen_without_limit === 'false',
+    );
+    const [proposedClosingTime, setProposedClosingTime] = useState('');
+
+    // We need to manage state on every change to the form.
+    useEffect(() => {
+      setShowTimeLimitOptions(
+        row.action === 'set_time_limit_all' || row.open || form.reopen_without_limit === 'false',
+      );
+      if (row.action !== 'set_time_limit_all') {
+        setProposedClosingTime(calculateClosingTime());
+      }
+    }, [form]);
+
+    // We need to manage state and update the form on each change. This function can
+    // be passed as a prop to each child componenent that needs to update the form.
+    const handleChange = (e: Event) => {
+      setForm({
+        ...form,
+        [(e.target as HTMLInputElement).name]: (e.target as HTMLInputElement).value,
+      });
+    };
```

Comment:
I had not considered type safety. I implemented the `updateFormState` function above in [27fb1ca](https://github.com/PrairieLearn/PrairieLearn/pull/10800/commits/27fb1ca4494de2dc7e45938cbadfd0923939c8f4) and now pass in the correct key/value for each event handler rather than picking it off of the event name.